File: pb-test/protobufv3_test/v3/src/encoder.cc
#include "msg.pb.h"
#include <fstream>
#include <string>
#include <stdio.h>

using namespace std;

int main(int argc, char** argv)
{
    ::proto3_proto::MessageArray ma;
    ::google::protobuf::Map< ::std::string, ::proto3_proto::Message>*
        msg_map_map = ma.mutable_msg_map();

    char name[64] = {0};

    for (int idx = 0; idx < 10; idx++) {
        sprintf(name, "%s%d", "hello", idx);
        ::proto3_proto::Message msg;
        msg.set_name(name);
        msg.set_hilarity(::proto3_proto::Message_Humour_PUNS);
        msg.set_height_in_cm(123456);
        msg.set_data("world");
        msg.set_result_count(123456);
        msg.set_true_scotsman(true);
        msg.set_score(123.456f);
        for (unsigned long index = 0; index < 96; index++) {
            msg.add_key(index);
        }

        printf("msg size:%d\n", msg.ByteSize());
        (*msg_map_map)[name] = msg;
    }

    printf("msg array size:%d\n", ma.ByteSize());

    fstream output("../pb_bin_v3", ios::out | ios::trunc | ios::binary);
    if (!ma.SerializeToOstream(&output)) {
        printf("Failed to write msg.\n");
        return -1;
    }

    google::protobuf::ShutdownProtobufLibrary();

    return 0;
}

File: pb-test/protobufv3_test/v3/go/encoder.go
package main

import (
	"fmt"
	"github.com/golang/protobuf/proto"
	"io/ioutil"
	"proto3_proto"
)

func main() {
	var ma proto3_proto.MessageArray
	ma.MsgMap = make(map[string]*proto3_proto.Message)
	for idx := 0; idx < 10; idx++ {
		var msg proto3_proto.Message
		msg.Name = fmt.Sprintf("%s%d", "hello", idx)
		msg.Hilarity = proto3_proto.Message_PUNS
		msg.HeightInCm = 123456
		msg.Data = []byte("world")
		msg.ResultCount = 123456
		msg.TrueScotsman = true
		msg.Score = 123.456
		msg.Key = make([]uint64, 100)
		for index := 0; index < 96; index++ {
			msg.Key[index] = uint64(index)
		}

		ma.MsgMap[msg.Name] = &msg
	}

	data, err := proto.Marshal(&ma)
	if err != nil {
		fmt.Println("marshaling error: ", err)
	}
	//fmt.Println("data:", string(data))

	ioutil.WriteFile("../../go_pb_bin_v3", data, 0666)
}
